### ISSUE TITLE
[LTX-2-DEV] Test of CREPA / DOP / prior_div / blank_pres

### ISSUE THREAD

**Issue by dorpxam** [LongCatFan1/longcat-tuner#28]
Not a full test, just a run for testing the new features.

## Workaround for WandB

<details>
<summary>Little bug on starting of training / WandB</summary>
<div>&nbsp;</div>

There's a little bug in `generate_step_logs` (hv_train_network.py / line 533).

```
if lr_descriptions is not None:
    lr_desc = lr_descriptions[i]
```

list out of range on : `lr_desc = lr_descriptions[i]`

Workaround:
```
if lr_descriptions is not None and i < len(lr_descriptions):
    lr_desc = lr_descriptions[i]
```

</details>

## Settings

First point, I will not able to use all the features (too heavy) even if I'm able to run with the settings:

```python
--flash_attn ^
--fp8_base ^
--fp8_scaled ^
--blocks_to_swap 36 ^
--use_pinned_memory_for_block_swap ^
--gradient_checkpointing ^
--gradient_accumulation_steps 1 ^
--optimizer_type AdamW ^
--learning_rate 2e-4 ^
--lr_scheduler constant_with_warmup ^
--lr_warmup_steps 300 ^
--network_module networks.lora_ltx2 ^
--network_dim 32 ^
--network_alpha 32 ^
--timestep_sampling shifted_logit_normal ^
```
And default settings for the new features:
```python
--blank_preservation ^
--blank_preservation_args multiplier=0.5 ^
--use_precached_preservation ^
--dop ^
--dop_args class=woman ^
--prior_divergence ^
--prior_divergence_args multiplier=0.1 ^
--crepa ^
--crepa_args mode=backbone student_block_idx=16 teacher_block_idx=32 lambda_crepa=0.1 tau=1.0 num_neighbors=2 schedule=constant warmup_steps=0 normalize=true
```

## Dataset / Cache

The dataset is my classic 'Twin Peaks' / 30 videos but this time with the use of full length for the shorts in low resolution:

```toml
[general]
batch_size = 1
enable_bucket = true
bucket_no_upscale = false

[[datasets]]
resolution = [320, 224]
video_jsonl_file = "E:/Training/ltx2-sherylin/dataset/dataset_vid_320.jsonl"
cache_directory = "E:/Training/ltx2-sherylin/dataset/cache_vid_320"
frame_extraction = "head"
target_frames = [1, 41, 49, 65, 73, 89, 105, 113, 121]
max_frames = 1441
source_fps = 24.0
target_fps = 24.0
num_repeats = 1

[[datasets]]
resolution = [640, 480]
video_jsonl_file = "E:/Training/ltx2-sherylin/dataset/dataset_vid_640.jsonl"
cache_directory = "E:/Training/ltx2-sherylin/dataset/cache_vid_640"
frame_extraction = "chunk"
target_frames = [25]
max_frames = 1441
source_fps = 24.0
target_fps = 24.0
num_repeats = 1
```
So:
- Head for each 30 videos in `320 x 224` resolution (1 + n-frames up to full length)
- Chunks of 1 second for each 30 videos in `640 x 480` resolution

The cache consist of 256 items:
- `cache_vid_320`: 174 items
- `cache_vid_640`: 82 items

Even if I do not use Dinov2 for CREPA, I've tested the caching too, so the full cache for latent (video/audio) + Dinov2 + Gemma:
- `cache_vid_320`: 20.70 GB
- `cache_vid_640`: 4.91 GB

## Training

I do not have train more than 1 of 12 epoch right now, the `s/it` is approximately `42 sec` in mean. So 3 hours for one epoch of 256 steps.
```
steps:   8%| 256/3072 [3:05:15<33:57:52, 43.42s/it, avr_loss=0.829, loss_a=0.67, loss_v=0.263]
```

The loss from WandB:

<img width="2506" height="1167" alt="Image" src="https://github.com/user-attachments/assets/5d795587-5f4b-4592-9e73-afdd25e82ce7" />

## Partial Result:

It's difficult to say if the result it's great or not after only one epoch but I can say:
- The training works
- The resulting LoRA (1 epoch) works too

## ComfyUI LoRA test

`LTX-2-Distilled-FP8 | 8 steps | 5 sec | 1536 x 864`

> A cinematic close-up of a young woman with a 1940s-inspired short brunette bob and elegantly arched eyebrows. She is wearing a thick, burgundy turtleneck sweater and sits in a nostalgic, dimly lit vintage diner. Initially, her head is tilted down, her focus entirely on her coffee cup as she rhythmically stirs it with a silver spoon. She then looks up with a sudden, expressive intensity, her eyes wide and bright, a playful yet knowing smirk forming on her lips as she addresses someone off-camera. The scene is shot on 35mm film with a warm, amber-toned color palette and a soft grain, capturing the cozy yet mysterious atmosphere of a classic 90s television drama. The rhythmic, crystalline clinking of a metal spoon against a porcelain cup dominates the foreground. This is layered over the soft, low-frequency hum of a quiet diner and the distant, muffled sound of a door opening. A young woman’s voice, soft and deliberate with a hint of breathiness, says: \"Agent Cooper loves coffee.\" The tone is intimate and slightly mischievous.

https://github.com/user-attachments/assets/f4ba1fe2-b0a2-4b91-a910-4824260335ad

`LTX-2-Distilled-FP8 | 8 steps | 5 sec | 1536 x 864`

> A cinematic close-up of a young woman with short, voluminous curly brunette hair, red lipstick, and bright eyes, wearing a white cardigan with a black tree-like pattern. The scene is bathed in warm, soft indoor lighting, creating a cozy and intimate atmosphere. She begins with a radiant, joyful smile, which transitions into a curious and polite expression as she turns her head slightly to the side. The audio consists of a soft, melodic instrumental track playing in the background, layered with the faint, ambient chatter of a public space. Her voice is gentle and inviting as she clearly asks, \"Can I sit here?\"

https://github.com/user-attachments/assets/5a467ef7-226a-4749-90c6-a7dd6dbe1fcf

## WELL DONE

Good job as always !!! Thanks a lot !!!

**Comment by AkaneTendo25** [comment]
Thanks for testing. DOP, BOP, and PD were introduced by Ostris, credit to him. Their code is relatively simple and compact, so porting them was not difficult.

As for CREPA, I doubt it will work out of the box, no tests were conducted yet🙂. I'm will test I2V training and plan to test CREPA with it as well.

After that, I'll probably merge `dev` into `ltx-2`, since most updates (except maybe FPS resampling) are opt-in and should not affect the existing training pipeline (unlike what `ltx-2-legacy` once did).

**Comment by dorpxam** [comment]
I have restart a quick LTX-2-DEV training on a bunch of 12 videos only (no audio / 352x192x297 frames) with the features:
- CREPA (backbone)
- Prior divergence
- 60 FPS

But I'm not sure about the about the negative loss/average & loss/current ? Is this a consequence of the negative essence of the loss/crepa & loss/prior_div (mean of the whole ?). I don't want to spend too much hours if there's something wrong. The 2nd epoch LoRA (rank-64) seem to work on ComfyUI without something strange, even if nothing is learn of course. For me it's OK, but can you confirm?

<img width="2500" height="1239" alt="Image" src="https://github.com/user-attachments/assets/e987c672-7916-40fe-af0c-de337f41e2b8" />

**Comment by dorpxam** [comment]
Yes. Works. 60FPS too. I can't produce sample because It's personal, but next training will be something like karate or maybe dancing test (high FPS). Anyway, I confirm that the (full) CREPA/High-FPS training works. Maybe 33% too much epochs for this case, but great !!!

<img width="2503" height="1278" alt="Image" src="https://github.com/user-attachments/assets/e63bfaf9-f558-4a76-b4fb-43e59c23337e" />

**Comment by dorpxam** [comment]
Interesting. By changing the CREPA default parameters (`--crepa_args`) by:

```ruby
mode = backbone 
student_block_idx = 16 
teacher_block_idx = 32 
lambda_crepa = 0.5 
tau = 1.0 
num_neighbors = 2 
schedule = cosine 
warmup_steps = 200 
normalize = true
```

The `loss/crepa` look different. Seem works when I look at the `loss/video` curve.

<img width="2503" height="1278" alt="Image" src="https://github.com/user-attachments/assets/8e1a29d6-c5bd-4c5e-b597-bece39f4a617" />

Honestly, I have no idea how to set the parameters, I have just adapt de T2V settings from SimpleTuner just for testing.

This one is a concept LoRA. At least I hope so.

- 12 videos [256x192x121] (so 2 seconds @ 60FPS)
- LoRA rank-64
- 60 FPS
- Prior Divergence
- CREPA

The dataset is very simple:

<img width="1330" height="369" alt="Image" src="https://github.com/user-attachments/assets/d4157deb-ecb2-468f-8caa-ccfb09f07151" />

The caption very basic too (just to see if the model will generalize with that kind of caption):

```jsonl
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/001.mp4", "caption": "metamorphosis between a man and a woman." }
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/002.mp4", "caption": "metamorphosis between a woman and a woman." }
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/003.mp4", "caption": "metamorphosis between a woman and a man." }
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/004.mp4", "caption": "metamorphosis between a man and a woman." }
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/005.mp4", "caption": "metamorphosis between a woman and a man." }
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/006.mp4", "caption": "metamorphosis between a man and a woman." }
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/007.mp4", "caption": "metamorphosis between a woman and a man." }
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/008.mp4", "caption": "metamorphosis between a man and a woman." }
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/009.mp4", "caption": "metamorphosis between a woman and a man." }
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/010.mp4", "caption": "metamorphosis between a man and a man." }
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/011.mp4", "caption": "metamorphosis between a man and a woman." }
{ "video_path": "E:/Training/ltx2-morphing/dataset/videos/012.mp4", "caption": "metamorphosis between a woman and a woman." }
```

**Comment by AkaneTendo25** [comment]
Curves look consistent, loss/current is video + crepa + prior_div so `~0.2 + (-3.0) + (~0) = -2.8.`
Video loss at `0.151` is very good, training is working. But CREPA is dominating total loss.
Your `lambda_crepa=0.5` is 5x the default, so CREPA gradient ends up like 15x stronger than actual denoising, and if your dataset is small, there isnt much temporal variation to learn so CREPA saturates by step ~400 and after that just drowns out the video signal without helping.

I'd drop `lambda_crepa` to `0.05-0.1`. And maybe cancel `prior_div` completely since we cant reliably "factorize" how CREPA and `prior_div` are working together.

**Comment by dorpxam** [comment]
Finally, here's the chart:

<img width="2500" height="1251" alt="Image" src="https://github.com/user-attachments/assets/c97b64c0-bcd3-4c7f-a340-a97a329b84d9" />

## THE RIGHT CHOOSE OF EPOCH

It's difficult to have a precise idea of the right epoch to select in the 48 epochs stored. Maybe, It will be cool to have a external script that allow a batch processing in inference for testing the N-epochs with same conditions (resolution/prompt/seed). Currently I do some test under ComfyUI but this is limited.

## CURRENT RESULT

I do a simple test with the prompt:

> metamorphosis between a man and a woman.

And do it conditioned by an audio for the fun. I stay in 768 x 512 | 60 FPS | 10 seconds. Higher resolution is more complex to produce here in 60 FPS.

#### TEST 1 : EPOCH 12 / STEP 576

This one fail to produce the woman metamorphosis but the hair move is cool.

https://github.com/user-attachments/assets/171018f7-c089-41b0-82d0-b1a5a89a790a

#### TEST 2 : EPOCH 24 / STEP 1152

Strangely, the epoch 24 fail to metamorphosis whatever the seed. Always that kind of guy.

https://github.com/user-attachments/assets/7bb9f50f-46b5-4a04-9747-de7c6fe43cb3

#### TEST 3 : EPOCH 32 / STEP 1536

Interesting. The model learn to produce slow metamorphosis. Trained on low-res (352 x 192) and with a metamorphosis approximately centered in the middle of the 2 seconds video. That's a quick change (1/4 maybe 1/2 second), but here in a 10 seconds video, that's smooth.

https://github.com/user-attachments/assets/588dee3c-0ec2-4f16-a2d8-bf63910872bb

#### TEST 4 : EPOCH 42 / STEP 2016

Probably a slightly overtrained, some test show this characteristic strange effect in the air. I'm sure that one of the others epoch works better. Maybe a more detailed prompt can avoid the blue uniform background.

https://github.com/user-attachments/assets/e397cf9b-dce1-42c5-be96-c4bbb7758d84

You can see the 'strange effect' I mean in this bad sample:

https://github.com/user-attachments/assets/8ea285c5-f54d-4a43-a42b-afe56508ba08

#### CONCLUSION

Globally, the LoRA is cool but too much fixed on the training data. I probably need a better caption to train. I've read some information that pretend the 'caption' is the description of what you don't want your LoRA to produce? 

For example, by writing "a blonde woman with pale skin", allow you to not fix the LoRA based on these characteristics. I really don't know. But it's clear that the LoRA produce characters that look like the training data (not precisely of course).

I start the same training without the prior divergence et with the corrected lambda for CREPA.

## EDIT

Not sure that 60 FPS LoRA works in other FPS, a test @ 30 FPS in ComfyUI produce a jerky effect, right? 

https://github.com/user-attachments/assets/dc68b686-eb56-47ec-beb7-b6fca408fb62

**Comment by AkaneTendo25** [comment]
@dorpxam Thanks! A lot of interesting observations. I don't have much free time this week, so I can't really devote enough time to LTX-2 trainer and your results right now, but I'll definitely check them later.
